packages/freezed/lib/src/templates/concrete_template.dart

```diff
@@ -370,25 +369,90 @@ class Property {
       );
     }
 
+    final decorators = parseDecorators(element.metadata);
+
     return Property(
       name: element.name,
       type: parseTypeSource(element),
-      decorators: parseDecorators(element.metadata),
+      decorators: parseDecorators(element.metadata).toList()
+        ..removeWhere((d) => d.contains('@JsonKey')),
       nullable: element.isNullable,
       defaultValueSource: defaultValue,
-      hasJsonKey: element.hasJsonKey,
+      jsonKeyOptions: JsonKeyOptions.fromDecorators(decorators),
     );
   }
 
   @override
   String toString() {
-    return '${decorators.join()} final $type $name;';
+    var jsonKeyOptions = this.jsonKeyOptions ?? JsonKeyOptions();
+
+    if (nullable) {
+      jsonKeyOptions = jsonKeyOptions.mergeValues({
+        'required': 'false',
+        'disallowNullValue': 'false',
+      });
+    } else {
+      jsonKeyOptions = jsonKeyOptions.mergeValues({
+        'required': 'true',
+        'disallowNullValue': 'true',
+      });
+    }
+
+    if (defaultValueSource != null) {
+      jsonKeyOptions = jsonKeyOptions.mergeValues({
+        'defaultValue': defaultValueSource,
+      });
+    }
+
+    return '${decorators.join()} $jsonKeyOptions final $type $name;';
   }
 
   Getter get getter => Getter(
       name: name, type: type, decorators: decorators, nullable: nullable);
 }
 
+class JsonKeyOptions {
+  final Map<String, String> values;
+
+  JsonKeyOptions([this.values = const {}]);
+
+  static JsonKeyOptions fromDecorators(List<String> decorators) {
+    if (decorators == null || decorators.isEmpty) return null;
+
+    String jsonKeyOptions;
+    for (final d in decorators) {
+      if (!d.startsWith('@JsonKey')) continue;
```

Comment:
There are extensions already for such thing. Search for the extensions on the project, you should find them

This shouldn't be done with a `.split`